File: java/StringCompression.java
public class StringCompression {
    public static String compress(String input) {
        if (input == null || input.isEmpty()) {
            return "";
        }

        StringBuilder compressed = new StringBuilder();
        int count = 1;

        for (int i = 1; i <= input.length(); i++) {
            if (i < input.length() && input.charAt(i) == input.charAt(i - 1)) {
                count++;
            } else {
                compressed.append(input.charAt(i - 1)).append(count);
                count = 1;
            }
        }

        return compressed.length() < input.length() ? compressed.toString() : input;
    }

    public static void main(String[] args) {
        String sample = args.length > 0 ? args[0] : "aabcccccaaa";
        System.out.println(compress(sample));
    }
}

File: kotlin/StringCompression.kt
fun compress(input: String): String {
    if (input.isEmpty()) return ""

    val compressed = StringBuilder()
    var count = 1

    for (i in 1..input.length) {
        if (i < input.length && input[i] == input[i - 1]) {
            count++
        } else {
            compressed.append(input[i - 1]).append(count)
            count = 1
        }
    }

    return if (compressed.length < input.length) compressed.toString() else input
}

fun main(args: Array<String>) {
    val sample = if (args.isNotEmpty()) args[0] else "aabcccccaaa"
    println(compress(sample))
}
